### PR TITLE
Improve sysproxy exception handling

### LINKED ISSUE #1487: 索引超出数组界限 - Sysproxy
Please read Wiki carefully, especially https://github.com/shadowsocks/shadowsocks-windows/wiki/Troubleshooting.
Please answer these questions before submitting your issue. Thanks!

### Version(release version or AppVeyor link)

4.0.6
### Environment(Operating system, .NET Framework, etc)

win10 
### Steps you have tried
run the client

### Config and error log in detail (with all sensitive info masked)
System.IndexOut

### PR DESCRIPTION
**Pull request by studentmain**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

- [x] [Searched](https://github.com/shadowsocks/shadowsocks-windows/search?q=is%3Apr&type=Issues) for similar pull requests
- [x] Compiled the code with Visual Studio

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

#1487
https://github.com/shadowsocks/shadowsocks-windows/issues/1991#issuecomment-422240535
#2013

- Sysproxy.exe sometimes output in UTF16LE, cause parse fail. Manually convert it and retry.
- when fail, log information (`sysproxy` args, `sysproxy query` return string), then clear proxy setting and retry if requested.

### CODE REVIEW

**Comment by studentmain** [comment]
鉴于#2013 ，还需要进一步调查

**Comment by studentmain** [comment]
一时半会没有进展，先关了吧

**Comment by celeron533** [comment]
暴力一点的做法：
接住异常，通知用户无法获取和备份当前代理设置，是否继续设置代理

**Comment by studentmain** [comment]
@celeron533 也行，顺便打个log收集故障详情。我再改改，好了再开。

File: shadowsocks-csharp/Data/ja.txt [hunk 1]
```diff
@@ -148,3 +148,6 @@ Proxy handshake failed=プロキシ ハンドシェイクに失敗しました
 Register hotkey failed=ホットキーの登錄に失敗しました。
 Cannot parse hotkey: {0}=ホットキーを解析できません: {0}
 Timeout is invalid, it should not exceed {0}=タイムアウト値が無効です。{0} 以下の値を指定して下さい。
+
+Error occured when process proxy setting, do you want reset current setting and retry?=处理代理设置时发生错误，是否重置当前代理设置并重试？
```
Comment (celeron533) [inline shadowsocks-csharp/Data/ja.txt]: Considering there is neither suitable Japanese nor Traditional Chinese translation, you may could remove these lines and just let the UI to display its original English text.

Comment (studentmain) [reply shadowsocks-csharp/Data/ja.txt]: I'll ask my friend for a translate, will update (or delete) it.

**Comment by studentmain** [comment]
@celeron533 感觉目前能做的就这些，看看还有什么要修的。

顺便一提，我认为问题的根源在于外置sysproxy且以纯文本方式交互。
- 要么重回wininet，不过那听上去不太对劲。
- 要么给sysproxy添加使用某种二进制信道与程序交互的能力。
  - 如果打算今后搞成sysproxy服务，开个管道或者socket应该不错
  - 要不就让sysproxy从文件中读写数据，这样也可以避免和终端打交道时出现的问题
- 要么干脆做成sysproxy.dll，这样P/Invoke的时候或许会舒服点，同时也可以更好地识别#2027 这种情况。

**Pull request merged**